dali/operators/reader/numpy_reader_op.cc

```diff
@@ -226,6 +234,92 @@ DALI_SCHEMA(NumpyReader)
 submodule and renamed to follow a common pattern. This is a placeholder operator with identical
 functionality to allow for backward compatibility.)code");  // Deprecated in 1.0;
 
+NumpyReaderCPU::~NumpyReaderCPU() {
+  // Stop the prefetch thread as it uses the thread pool from this class. So before we can
+  // destroy the thread pool make sure no one is using it anymore.
+  this->StopPrefetchThread();
+}
+
+void NumpyReaderCPU::Prefetch() {
+  // We actually prepare the next batch
+  DomainTimeRange tr("[DALI][NumpyReaderCPU] Prefetch #" + to_string(curr_batch_producer_),
+                      DomainTimeRange::kRed);
+  DataReader<CPUBackend, NumpyFileWrapper>::Prefetch();
+
+  if (!dont_use_mmap_)
+    return;
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+
+  for (auto &target : curr_batch) {
+    if (target->data.shares_data()) {
+      target->data.Reset();
+    }
+    if (use_o_direct_) {
+      /*
+       * the offset to the data we are interested in
+       *                  |<-                       aligned_len                    ->|
+       *                  |                                                          |
+       *                  |                               |<- target->nbytes ->|     |
+       *                  |<-    target_data_offset     ->|                    |     |
+       *     |<-          target->data_offset           ->|
+       *     -------------********************************XXXXXXXXXXXXXXXXXXXXXX******
+       *     |            |                               |                          |
+       * file_start   block_start/allocated_memory     shared_ptr                   block_end
+       */
+
+      // align the size of data to read accounting for its start
+      auto aligned_len = align_up_offset(target->nbytes, target->data_offset,
+                                         o_direct_read_len_alignm_);
```

Comment:
This doesn't seem right.
Let's assume:
alignment = 4k
data_offset = 4k - 10
nbytes = 1
Now, we have `align_up_offset(1, 4086) == align_up(1 + align_remainder(4086, 4096), 4096) == align_up(4087, 4096) == 4096`
That's incorrect, because we'll actually have to read two pages.
I think it should rather be calculated as:
```
block_start = align_down(target->data_offset, alignment)
block_end = align_up(target->data_offset + target->nbytes, alignment)
aligned_len = block_end - block_start
```

Reply:
Done